File: 124.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static int ans = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        if (root == null){
            return 0;
        }   
        maxPathSumHelper(root);
        int res = ans;
        ans = Integer.MIN_VALUE;
        return res;
    }
    
    private static int maxPathSumHelper(TreeNode nd){
        if (nd == null){return 0;}
        int l = Math.max(0, maxPathSumHelper(nd.left));
        int r = Math.max(0, maxPathSumHelper(nd.right));
        int sum = l+r+nd.val;
        ans = Math.max(ans, sum);
        return Math.max(l,r)+nd.val;
    }
}

File: 124.py
#124. binary tree maximum path sum
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        def maxGain(nd):
            nonlocal maxSum
            if not nd:
                return 0
            
            leftmg = max(maxGain(nd.left), 0)
            rightmg = max(maxGain(nd.right), 0)
            
            price_newpath = leftmg+rightmg+nd.val
            maxSum = max(maxSum, price_newpath)
            
            return nd.val + max(leftmg, rightmg)
        
        maxSum = float('-inf')
        maxGain(root)
        return maxSum